Design of data preprocessing in AutoML.md

```diff
@@ -0,0 +1,93 @@
+
+PipelineModel model could be seen as a proxy to a main scoring model. Even name on leaderboard should reflect this fact - keep original name of model (maybe just add suffix "_pipeline" that signifies that preprocessing was used).
+
+#### Pros:
+-  will contribute to a simmetry between native and mojo/production aspects of the codebase. In production we use MojoPipeline to fullfil corresponding task
+- will probably provide better isolation in terms of how many classes we will have to "touch" or affect
+- as pipeline is a separate concept it will be a place for doing pipeline related logic, e.g. how sequence of the preprocessing models is going to be applied, how data flows from preprocessor to preprocessor( input and outputs). I believe that this functionality is not Model specific and could/should be abstracted out.
+
+#### Cons:
+- wrapping original main model into PipelineModel will hide its functionality. It can be mitigated by introducing ProxyModel interface with `getMainModel` method that will provide main model to the user. Though, for the clients (theoretically) PipelineModel could be flattenned automatically. Complexity of doing this  depends on how flexible we are with serialisation at rest api layer.
+
+![PipelineModel approach](pipelinemodel.png)
+
+(2) *Integration preprocessing functionality into a Model class*
+
+We can probably have two sub-approaches here:
+- when we have PipelineModelBuilder and during training it collapses into main model without any wrapper around it
+- when we extend functionality of ModelBuilder by adding ability to have references to proprocessing models
+
+Not sure which approach is better but probably we can avoid introduction of the PipelineModelBuilder class. But it will increase number of responsibilities of the ModelBuilder class.
+
+#### Pros:
+- We do not hide any functionality behind PipelineModel wrapper
+- All models could be equally seen as pipelines ( which could also be a cons.)
+
+#### Cons:
+- conceptually abusing notion of the Model. Model ( what is also known as Transformer in other frameworks) is supposed to be only trained on data and after that be able to transform similar data. Nothing about data preprocessing is expected here. See Concerns section for more details.
+- new functionality will be exposed not only to AutoML scope. All algos will be "aware" about data preprocessing from now on. With first approach we would just add notion of ability to be a Proxy model and not all the models will have to realise this interface. Though, absense of the preprocessing models could be seen as "no-impact" on Model, still from code perspective there will a bigger impact. Hopefully preprocessing functionality will be generic enough without depenedencies on a particular types of the models.
```

Comment:
> new functionality will be exposed not only to AutoML scope

All algos are aware of Categorical Encodings already, I don't see it as an issue. Also, it doesn't need to increase complexity of the Model class at all. The implementation could simply allow passing an external encoder (interface) instead of coupling the Model with Target Encoder capabilities. This could be available to both Java/Python/R.

Reply:
I was talking in terms of new TE functionality. Even though Models, as you just have said, are aware of Categorical Encoding, they will not be aware of TE in case we use PipelineModel. So I think my point still valid.

But agree, knowing that Models already have embedded preprocessing logic like Categorical Encodings means we can add TE in the same fashion and it will not increase complexity in terms of types of responsibilities ( as we already have preprocessing type of activity)